src/coreclr/vm/methodtable.cpp

```diff
@@ -4222,9 +4223,20 @@ OBJECTREF MethodTable::GetManagedClassObject()
 
         REFLECTCLASSBASEREF  refClass = NULL;
         GCPROTECT_BEGIN(refClass);
-        refClass = (REFLECTCLASSBASEREF) AllocateObject(g_pRuntimeTypeClass);
 
-        LoaderAllocator *pLoaderAllocator = GetLoaderAllocator();
+        LoaderAllocator* pLoaderAllocator = GetLoaderAllocator();
+        if (!pLoaderAllocator->CanUnload())
+        {
+            // Allocate RuntimeType on a frozen segment
+            FrozenObjectHeapManager* foh = SystemDomain::GetFrozenObjectHeapManager();
+            size_t objSize = g_pRuntimeTypeClass->GetBaseSize();
+            refClass = (REFLECTCLASSBASEREF)ObjectToOBJECTREF(foh->TryAllocateObject(g_pRuntimeTypeClass, objSize));
```

Comment:
I think this memory is leaked if the InterlockedCompareExchangeT below fails. Might be okay.

Reply:
It may be a good idea to just take a lock around this so that there is no leak.

TryAllocateObject may take a lock, and so you can end up in a pathological situations where a bunch of threads get stacked up behind this lock and all of them leak.

This leak is not a problem for non-frozen instances. The GC will clean them up.